packages/teleport/src/services/websession/storeWebSession.ts

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import history from 'teleport/services/history';
+
+import {
+  KeysEnum,
+  EventKeys,
+  BroadcastChannelMessage,
+  BearerToken,
+} from './types';
+
+const isTest = process.env.NODE_ENV === 'test';
+
+export class StoreWebSession {
+  bcBroadcaster: BroadcastChannel;
+  bcReceiver: BroadcastChannel;
+
+  constructor() {
+    this.initBroadcastChannel();
+  }
+
+  clear() {
+    this.bcBroadcaster.postMessage({ key: EventKeys.CLEAR });
+    history.goToLogin();
+  }
+
+  setBearerToken(token: BearerToken) {
+    this.bcBroadcaster.postMessage({ key: EventKeys.UPSERT_TOKEN, token });
+  }
+
+  setIsRenewing(isTokenRenewing: boolean) {
+    this.bcBroadcaster.postMessage({
+      key: EventKeys.UPSERT_TOKEN_IS_RENEWING,
+      isTokenRenewing,
+    });
+  }
+
+  setLastActive(lastActive: number) {
+    this.bcBroadcaster.postMessage({
+      key: EventKeys.UPSERT_LAST_ACTIVE,
+      lastActive,
+    });
+  }
+
+  getIsRenewing(): boolean {
+    return window.sessionStorage.getItem(KeysEnum.TOKEN_RENEW) === 'true';
+  }
+
+  getLastActive() {
+    const time = Number(window.sessionStorage.getItem(KeysEnum.LAST_ACTIVE));
+    return time ? time : 0;
+  }
+
+  getAccessToken() {
+    const bearerToken = getBearerToken();
+    return bearerToken ? bearerToken.accessToken : null;
+  }
+
+  getSessionInactivityTimeout() {
+    const bearerToken = getBearerToken();
+    const time = Number(bearerToken.sessionInactiveTimeout);
+    return time ? time : 0;
+  }
+
+  initBroadcastChannel() {
+    this.bcBroadcaster = new BroadcastChannel('websession_store');
+    this.bcReceiver = new BroadcastChannel('websession_store');
+
+    this.bcReceiver.onmessage = ({
```

Comment:
To support passing in a custom configuration this can be moved to a static method `attachBroadcastHandlers`. So  in a test this class would be called using something like...

```ts
const broadcastChannel = new BroadcastChannel('websession_store');
StoreWebSession.attachBroadcastHandlers(broadcastChannel)
const webSession = new StoreWebSession({broadcastChannel});
// setup initial messages for test
webSession.bcBroadcaster.postMessage({
  key: EventKeys.GET_TOKEN,
});
```

Reply:
A static method wouldn't do anything to `webSession` in this context, I don't think `attachBroadcastHandlers` is needed - I agree that it should be accepted in the constructor for mocking during tests, however.

Reply:
Oops sorry I had a typo in my pseudo code :D (updated). Yeah it's entirely possible that it's not required. I just wanted to illustrate a way that similar handlers could be added outside of the class and passed in. Lots of different ways to accomplish this.